### PR TITLE
Refactor: Utilizes function to load animations

### PR DESCRIPTION
**Pull request by onurtore**
Utilizes new LoadAnimations function to load animations trajectories during actor creation.

Part of three PR for the https://github.com/gazebosim/gz-sim/issues/1564#issue-1290407501 for slimming down the CreateActor function.

### CODE REVIEW

**Comment by onurtore** [comment]
@ahcorde, 
Sorry for the old version, this must be a better.

File: src/rendering/SceneManager.cc [hunk 1]
```diff
@@ -2327,3 +2226,144 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
   animData.valid = true;
   return animData;
 }
+
+std::unordered_map<std::string, unsigned int>
+SceneManager::LoadAnimations(const sdf::Actor &_actor)
+{
+
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2233]:
```suggestion
```

File: src/rendering/SceneManager.cc [hunk 2]
```diff
@@ -2327,3 +2226,144 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+  for (unsigned i = 0; i < _actor.AnimationCount(); ++i)
+  {
+    std::string animName = _actor.AnimationByIndex(i)->Name();
+
+    std::string animFilename = asFullPath(
+        _actor.AnimationByIndex(i)->Filename(),
+        _actor.AnimationByIndex(i)->FilePath());
+
+    gzdbg << "AnimFilename " << animFilename << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2252]:
```suggestion
    gzdbg << "Loading actor [" << actor.Name() << " ] animation filename [" << animFilename << "]" << std::endl;
```

File: src/rendering/SceneManager.cc [hunk 3]
```diff
@@ -2327,3 +2226,144 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+
+    double animScale = _actor.AnimationByIndex(i)->Scale();
+
+    gzdbg << "AnimScale " << animScale << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2256]:
```suggestion
    gzdbg << "Actor [" << actor.Name() << " ] animimation scale [" << animScale << "]" << std::endl;
```

File: src/rendering/SceneManager.cc [hunk 4]
```diff
@@ -2327,3 +2226,144 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+
+    std::string extension = animFilename.substr(animFilename.rfind('.') + 1,
+                                  animFilename.size());
+    std::transform(extension.begin(), extension.end(),
+                    extension.begin(), ::tolower);
+
+    gzdbg << "Extension " << extension << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2263]: is this debug trace required?

File: src/rendering/SceneManager.cc [hunk 5]
```diff
@@ -2327,3 +2226,144 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+        }
+      }
+      animMesh = meshManager->MeshByName(animFilename);
+
+      // add the first animation
+      auto firstAnim = animMesh->MeshSkeleton()->Animation(0);
+      if (nullptr == firstAnim)
+      {
+        gzerr << "Failed to get animations from [" << animFilename
+                << "]" << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2303]:
```suggestion
        gzerr << "Failed to get animations from [" << animFilename
              << "]" << std::endl;
```

File: src/rendering/SceneManager.cc [hunk 6]
```diff
@@ -2327,3 +2226,144 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+        std::unordered_map<std::string, unsigned int> emptyMap;
+        return emptyMap;
+      }
+      // do not add duplicate animation
+      // start checking from index 1 since index 0 is reserved by skin mesh
+      bool addAnim = true;
+      gzdbg << "Remember animName " << animName << std::endl;
+      gzdbg << "Number of animations node " << meshSkel->AnimationCount() << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2311]: remove ?

File: src/rendering/SceneManager.cc [hunk 7]
```diff
@@ -2327,3 +2226,144 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+        gzdbg << "\tAnimation name " << meshSkel->Animation(a)->Name() << std::endl;
+        if (meshSkel->Animation(a)->Name() == animName)
+        {
+          addAnim = false;
+          break;
+        }
+      }
+      if (addAnim)
+      {
+        gzdbg << "Adding Animation" << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2324]:
```suggestion
        gzdbg << "\t Adding Animation" << std::endl;
```

File: src/rendering/SceneManager.cc [hunk 8]
```diff
@@ -2327,3 +2226,144 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+          {
+            std::pair<double, math::Matrix4d> keyFrame = nodeAnim->KeyFrame(k);
+            skelAnim->AddKeyFrame(
+                node->Name(), keyFrame.first, keyFrame.second);
+          }
+        }
+
+        meshSkel->AddAnimation(skelAnim);
+      }
+      gzdbg << "Adding animation is done" << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2364]:
```suggestion
      gzdbg << "Animation was added << std::endl;
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Review by ahcorde — CHANGES_REQUESTED** [review]
linters are also failing 

```
   /github/workspace/src/rendering/SceneManager.cc:2311:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/rendering/SceneManager.cc:2312:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
  /github/workspace/src/rendering/SceneManager.cc:2315:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/rendering/SceneManager.cc:2346:  Should have a space between // and comment  [whitespace/comments] [4]
  /github/workspace/src/rendering/SceneManager.cc:2352:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/rendering/SceneManager.cc:2352:  Should have a space between // and comment  [whitespace/comments] [4]
```

File: src/rendering/SceneManager.cc (hunk 4 above)
Comment (onurtore) [reply src/rendering/SceneManager.cc]: No, none of these should be here.

File: src/rendering/SceneManager.cc [hunk 9]
```diff
@@ -2327,3 +2226,144 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+        // to meshSkel, changing the name that would also change the name of
+        // other instances of the animation
+        // todo(anyone) cloning is inefficient and error-prone. We should
+        // add a copy constructor to animation classes in ign-common.
+        // The proper fix is probably to update ign-rendering to allow it to
+        // load multiple animations of the same name
+        common::SkeletonAnimation *skelAnim =
+            new common::SkeletonAnimation(animName);
+
+        gzdbg << "Node Count " << meshSkel->NodeCount() <<std::endl;
```
[the remark this answers is not in the document]
Comment (onurtore) [reply src/rendering/SceneManager.cc]: No, none of these should be here.

File: src/rendering/SceneManager.cc [hunk 10]
```diff
@@ -2327,3 +2226,120 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+    std::string animFilename = asFullPath(
+        _actor.AnimationByIndex(i)->Filename(),
+        _actor.AnimationByIndex(i)->FilePath());
+
+    double animScale = _actor.AnimationByIndex(i)->Scale();
+
+    std::string extension = animFilename.substr(animFilename.rfind('.') + 1,
+                                  animFilename.size());
+    std::transform(extension.begin(), extension.end(),
+                    extension.begin(), ::tolower);
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2255]:
```suggestion
                   extension.begin(), ::tolower);
```

File: src/rendering/SceneManager.cc [hunk 11]
```diff
@@ -2327,3 +2226,120 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+      // Load the mesh if it has not been loaded before
+      const common::Mesh *animMesh = nullptr;
+      if (!meshManager->HasMesh(animFilename))
+      {
+        animMesh = meshManager->Load(animFilename);
+        if (animMesh->MeshSkeleton()->AnimationCount() > 1)
+        {
+          ignwarn << "File [" << animFilename
+              << "] has more than one animation, but only the 1st one is used."
+              << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2285]:
```suggestion
          ignwarn << "File [" << animFilename
                  << "] has more than one animation, but only the 1st one is used."
                  << std::endl;
```

File: src/rendering/SceneManager.cc [hunk 12]
```diff
@@ -2327,3 +2226,120 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+        }
+      }
+      animMesh = meshManager->MeshByName(animFilename);
+
+      // add the first animation
+      auto firstAnim = animMesh->MeshSkeleton()->Animation(0);
+      if (nullptr == firstAnim)
+      {
+        ignerr << "Failed to get animations from [" << animFilename
+                << "]" << std::endl;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2295]:
```suggestion
        ignerr << "Failed to get animations from [" << animFilename
               << "]" << std::endl;
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Review by ahcorde — CHANGES_REQUESTED** [review]
```
   /github/workspace/include/ignition/gazebo/rendering/SceneManager.hh:206:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
  /github/workspace/include/ignition/gazebo/rendering/SceneManager.hh:206:  Add #include <unordered_map> for unordered_map<>  [build/include_what_you_use] [4]
  /github/workspace/src/rendering/SceneManager.cc:2284:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
https://github.com/gazebosim/gz-sim/runs/7138943766?check_suite_focus=true

File: src/rendering/SceneManager.cc [hunk 13]
```diff
@@ -2327,3 +2226,121 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+      }
+      animMesh = meshManager->MeshByName(animFilename);
+
+      // add the first animation
+      auto firstAnim = animMesh->MeshSkeleton()->Animation(0);
+      if (nullptr == firstAnim)
+      {
+        ignerr << "Failed to get animations from [" << animFilename
+               << "]" << std::endl;
+        return nullptr;
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2297]: I'm not really sure if this should return `{}` or `mapAnimNameId`

**Review by ahcorde — CHANGES_REQUESTED** [review]
it's not compiling, see the comment

https://github.com/gazebosim/gz-sim/runs/7139388778?check_suite_focus=true

**Review by ahcorde — CHANGES_REQUESTED** [review]
there are some conflicts too

**Comment by onurtore** [comment]
> there are some conflicts too

Fixed the compilation error, did not see any other problems in the Build. Am i missing somewhere?

File: include/ignition/gazebo/rendering/SceneManager.hh [hunk 14]
```diff
@@ -23,6 +23,7 @@
 #include <string>
 #include <utility>
 #include <vector>
+#include <unordered_map>
```
Comment (ahcorde) [inline include/ignition/gazebo/rendering/SceneManager.hh:26]: alphabetize headers

File: src/rendering/SceneManager.cc [hunk 15]
```diff
@@ -2327,3 +2225,124 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+  common::SkeletonPtr meshSkel = descriptor.mesh->MeshSkeleton();
+
+  unsigned int numAnims = 0;
+  std::unordered_map<std::string, unsigned int> mapAnimNameId;
+  mapAnimNameId[descriptor.meshName] = numAnims++;
+
+  // Load all animations
+  for (unsigned i = 0; i < _actor.AnimationCount(); ++i)
+  {
+
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2245]:
```suggestion
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
There are some conflicts with `ign-gazebo6` too

File: src/rendering/SceneManager.cc (hunk 15 above)
Comment (onurtore) [reply src/rendering/SceneManager.cc]: Done.

**Review by ahcorde — APPROVED** [review]

**Pull request merged**